### ISSUE TITLE
Make Nominatim installable

### ISSUE THREAD

**Issue by lonvia** [openstreetmap/Nominatim#634]
This issue tracks requirements and progress for creating a proper 'make install ' target. 

CC @melvyn-sopacua 

## Current situation

Since the switch to cmake, all code is already built in a separate build directory. Nominatim is then expected to be run from the build directory but also uses some files from the source directory. The cmak process configures the paths in the settings.php file to point to source (CONST_BasePath) and the build directory (CONST_InstallPath). All files are accessed relative to one of these two constants.

## Installable version

The files used after building are

 * from the source directory
   * `data/*`
   * `lib/*`
   * `munin/*` - not really required but useful to have around
   * `sql/*`
   * `utils/tigerAddressImport.py`
 * from the build directory
   * `docs/*` - useful to have
   * `module/nominatim.so`
   * `nominatim/nominatim`
   * `osm2pgsql/osm2pgsql`
   * `utils/*.php` - some are helper scripts for development that might be left out for installation
   * `website/*`

There are a few requirements for the installation process:

 * Nominatim must remain executable in the build directory to make development easier.
 * osm2pgsql must always be installed in the exact version provided as there may be interdependencies between osm2pgsql and Nominatim source code.

There is no problem with reorganizing files and directories if it makes the installation process easier.

### Php code

Currently part of the php code (in website and util) is copied into the build directory to be able to modify some paths. That is not ideal as cmake needs to be rerun on every modification. Ideally the build directory should just contain thin wrappers that call the source in the source directory/the installed php code.

### Database-dependent configuration directories

There are some files which are closely tied to a given database import. An obvious candidate is `settings/local.php`. In the future I'd also like to copy part of the code (SQL and PHP) which is tied to the database content into a dedicated directory during import, so that it becomes easier to update the software without reimporting the entire database. See https://github.com/lonvia/Nominatim/tree/site-installation for a branch that starts splitting off database-dependent code.

This is not directly related to the installation issue but something to take into account when reorganising the include paths.h

**Comment by melvyn-sopacua** [comment]
One issue is how included files are referenced:

```php
require_once '../../lib/file.php';
```
This will include `file.php` two directories up **from whatever is the current working directory**.
```php
require_once dirname(__FILE__).'/../../lib/file.php');
```
This will always reference the same file, two directories up **from the current file**. IDE's like phpstorm have a little trouble with the second, but that's not a concern.
One area this was exposed, is the test runner for phpunit: if one isn't in the right directory, it fails horribly.

**Comment by lonvia** [comment]
We have been using the second version in most scripts. The tests are an oversight.

**Comment by melvyn-sopacua** [comment]
```sh
diff -ur /usr/obj/ports/usr/ports/astro/nominatim-devel/work/Nominatim-v2.6.0a2/utils /usr/obj/ports/usr/ports/astro/nominatim-devel/work/.build/utils/
Only in /usr/obj/ports/usr/ports/astro/nominatim-devel/work/Nominatim-v2.6.0a2/utils: import_wikipedia.sh
Only in /usr/obj/ports/usr/ports/astro/nominatim-devel/work/Nominatim-v2.6.0a2/utils: tigerAddressImport.py
```
Not surprising, cause there's nothing to configure in utils/*.php. Work in progress or oversight?

**Comment by melvyn-sopacua** [comment]
Update:
Got a preliminary install mechanism in place using FreeBSD port's install facilities. Going to test if that runs and if not, investigate. Once that's running, I'll have a good idea how to handle cmake.

**Comment by lonvia** [comment]
`import_wikipedia.sh` is only used for development, not something to distribute. `tigerAddressImport.py` I forgot. I've amended the list above.

**Comment by melvyn-sopacua** [comment]
Normally, you embed `@VARIABLE_NAME@` in a "configured file", but these don't contain any, so they're identical to their source counterpart. From the looks of it, the reason they're copied to the build directory is because they need access to the settings.

Are you open to using [PHP dot env](https://github.com/vlucas/phpdotenv) to point to the settings file? I would also migrate the "debug" setting to this, so it's easy to deploy test / production systems without mangling source files.

**Comment by lonvia** [comment]
I see, it seems I've misunderstood your diff question. You were asking why the php files are copied verbatim. Indeed, that is because they expect to find the configured settings file at a fixed relative path. It's all rather hacky, I'm open to changing that.

Could you elaborate how you would want to use the .env file? (Or do a quick proof-of-concept PR if that is faster.) Thinking about all that.. it would also be nice to keep the settings/local.php out of the build and source directory during development, so that is possible to do a quick `rm -rf`without loosing settings. Sounds like the env stuff could solve that, too.

**Comment by melvyn-sopacua** [comment]
In short, an application's behavior is configured by 2 different types of settings:

- Business logic settings: controlling features of the application
- Machine / Site dependant settings: controlling how to interact with services it needs

These are typically different roles as well: systems / network admins decide what you can use and where it's at, while features are decided by the user of the application.
Since Nominatim interacts with a bunch of external services, I want to tread carefully, but in the end, you'd want anything referencing something external to be in the env file. Laravel has a [nice example](https://github.com/laravel/laravel/blob/master/config/database.php).
I'll whip something up, but not a lot of time tomorrow, so probably thursday.

**Comment by melvyn-sopacua** [comment]
Work in progress:
https://github.com/melvyn-sopacua/Nominatim/compare/phpdotenv-integration?expand=1

**Comment by mtmail** [comment]
I'm quite familiar with composer and dotenv and used it a lot in other projects. In my opinion it's the right step forward to make Nominatim more 'modern'. This is great. On the other hand I'm not familiar with cmake so I can't comment on the '/lib' directory.

As you've noticed the Nominatim repository is quite a mix of PHP, C++, Python. When I did the travis-ci setup I wasn't even sure what the main programming language is (turns out setting it to Python made the travis-ci installation faster).

**Comment by melvyn-sopacua** [comment]
Thank you.
Even if you're not into the latest fashion, there's a few advantages for those unfamiliar with it:

- We can basically bring down php dependencies to "composer" and whatever core extension we may use. The rest is dictated by `composer.json`.
- We can migrate to using the phpunit in `vendor/bin` for testing
- You can control versions of packages eliminating errors due to version conflicts
- Autoloading

As for becoming more modern, my hands are itching and it's hard to control myself and not make this a refactoring in the process. I see methods that span >900 lines, PEAR-DB is pretty much dead and replaced by Pear-MDB2, methods use global variables that are only set because other files are included in the file that uses the class. Anyway - not out to trash past work, but there's a lot that can be improved.
As far as I can tell, Pear is only used for DB abstraction, so switching to PDO_pgsql would be my first step forward and probably something I'll work on after this, if only to improve my knowledge of the software.

**Comment by lonvia** [comment]
Wait, slow down. please. Introducing php package managers is an entirely different discussion (and not one I'm very favourable about at the moment, I'm afraid to say), so if you are interested in adding composer, please open a different ticket for discussion.

Some comments to the code above:

There does not seem to be much of an advantage of phpdot to the current local.php based system. We loose typed variables (env is string only, I suspect) and add the burden to have to add each variable twice. 

The COPY_TREE part seems to do exactly the opposite of what I outlined above. If I understand right it copies the entire php code in the build tree. So, the net result would be that a developer has to run cmake every time they change something in the php code. My requirement was that I basically never have to run cmake unless there is a fundamental change to the code structure. That means: as far as possible php code is used from the source code directly when running from the build directory. Only when somebody runs 'make install' should it be copied from the source directory to the final destination.

On a more general note: please try to focus on one problem at the time. There is some potentially good stuff in there (e.g. the HAVE_PROC_CPUINFO check) which is completely unrelated. I know it is easy to get sidetracked when there is so much to improve but it makes it considerably more difficult to review the code when you just have to mentally sort out which code belongs where.

**Comment by melvyn-sopacua** [comment]
I understand your concerns, but part of them are because I'm trying to transition slowly. For example, because the configuration uses constants, I'm now fetching them in place. Normally, you'd not use constants but fetch them from env directly, but that would require rewriting half of the code.

You're right though about the scope creep. So I'll factor out the minimally needed things, drop dotenv from this issue and prepare a new branch.

However, if you want to move forward to testable, maintainable php code, using tools and patterns familiar to php developers, then there will be rewrites that are not small manageable bits or at best won't deliver the promise until other bits are done as well.

**Comment by melvyn-sopacua** [comment]
Two questions:
CMakeLists.txt: 8 space indents / tabs / 4 space indents?

> Currently part of the php code (in website and util) is copied into the build directory to be able to modify some paths.
Could you point out those bits? I can't find any reference to `fwrite`, `file_put_contents` or even `fopen` in website. Only in utils/setup.php and utils/update.php which I think are run in command line, not website.

**Comment by lonvia** [comment]
4 spaces, see CONTRIBUTING.md.

The files are copied [here](https://github.com/twain47/Nominatim/blob/master/CMakeLists.txt#L102). They are not modified but each of the copied files has a line

    require_once(dirname(dirname(__FILE__)).'/settings/settings.php');

That means they refer to a settings file relative to their absolute position. The settings file itself is the one which is modified on installation. So the only reason to copy them is to get them at the right place to be able to include the settings from the build directory.

Better would be something like this: instead of copying the entire file, just create a file in the build directory which contains basically two lines:

   require_once(dirname(dirname(__FILE__)).'/settings/settings.php');
   require(@CMAKE_SOURCE_DIR@/website/whatever.php);

**Comment by melvyn-sopacua** [comment]
Why not trust include_path and use relative reference?

**Comment by lonvia** [comment]
How would you point it to the build directory?

**Comment by melvyn-sopacua** [comment]
Like any other php settings: php.ini, .htaccess, .user.ini or webserver config. You have to restart webserver if build directory changes location and this is the same requirement for the include path setting.

**Comment by lonvia** [comment]
It gets complicated when working with multiple build directories in parallel. I admit that I'm lazy and just want to do `cd build-someoddthing; ./utils/setup`. But maybe that's not possible to keep anyways when going to database-dependent directories.

**Comment by melvyn-sopacua** [comment]
Sadly, I know this will become easier when I switch configuration to dotenv the way I'd like to do it, but you won't believe me till it's finished :).
There won't be a settings.php anymore then, .env would be the settings. So cd into whatever build directory, would automagically have the right settings.

**Comment by lonvia** [comment]
So the net result would be: `cd build-someoddthing; ../utils/setup.php` ? That would be fine.

**Comment by balajiputhiran** [comment]
i to have same problem while configure after Make doesn't work. How to resolve the issues 

[ 74%] Built target osm2pgsql_lib
Linking CXX executable osm2pgsql
/usr/bin/ld: libosm2pgsql.a(reprojection.cpp.o): undefined reference to symbol 'log@@GLIBC_2.2.5'
//lib/x86_64-linux-gnu/libm.so.6: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
make[2]: *** [osm2pgsql/osm2pgsql] Error 1
make[1]: *** [osm2pgsql/CMakeFiles/osm2pgsql.dir/all] Error 2
make: *** [all] Error 2

**Comment by lonvia** [comment]
Please don't post on unrelated issues and do not cross -post your issues.

**Comment by sporkman** [comment]
So I see some talk of a FreeBSD port in this thread.  Has that happened or no?  I'm sitting here with an install that has probably failed in some portion of setup.php that's not logged an error. I have an empty search_name table and it's maddening.

**Comment by lonvia** [comment]
Please don't comment on random issues. This has nothing to do with FreeBSD installation.

**Comment by sporkman** [comment]
See https://github.com/openstreetmap/Nominatim/issues/634#issuecomment-281214791

That's clearly a fragment of a FreeBSD port in progress.

**Comment by lonvia** [comment]
You completely missed the point. If you want to get help, then please respect the most basic rules of etiquette: a) don't cross post your problems on multiple forums, b) don't add comments on random tickets that just happen to contain a keyword that might or might not be the cause of your problems, c) don't post on closed tickets, d) don't expect to this issue tracker to fill your gaps of knowledge in system administration.

**Comment by mtmail** [comment]
I'm closing this issue now. The discussions diverged too much to unrelated issues. Anybody finding this page via a websearch: just open a new issue.

Having this issue open for a year caused some confusion later and the title "Make Nominatim installable" didn't help. It was always somewhat installable and stable, just with the cmake approach there's room to improve file paths. Some work already found its way into Nominatim 3.0 and 3.1. A lot of lonvia's site-installation branch was also split off and merged via different PRs in the meantime.

melvyn-sopacua offered help on PHP .env files, composer package manager, database abstraction. Back in 2017 we already introduced too many changes in the installation process for the next release. Those are now covered in https://github.com/openstreetmap/Nominatim/milestone/2 I think him using FreeBSD was just coincidence, happy to look at PRs showing how the current (Nominatim 3.1) installation would look like on FreeBSD.

@lonvia Having the same files, e.g. `utils/setup.php`, in two directories (CONST_BasePath, CONST_InstallPath) still confuses some users. It is worth creating a new issue with almost the same description but different title?
